From ee34d49a96f7a9dfb15abf939348e848bb5799e3 Mon Sep 17 00:00:00 2001
From: Thu Nguyen <thu@os.amperecomputing.com>
Date: Mon, 21 Jul 2025 02:41:59 +0000
Subject: [PATCH] redfish_host_interface: Support GetBootStrapAccount

Support `GetBootStrapAccounnt` Ipmi command as section `8.1.2 Get
bootstrap account credentials (NetFn 2Ch, Command 02h)` in DSP0270
V1.3.1. The code will use OpenSSL APIs to create one random 16
characters username and one random 16 characters password to response
for command `GetBootStrapAccounnt`.

The username policy:
+ Include 16 characters (Counted Null Character)
+ Must include at least one low case character
+ Must include at least one up case character
+ Must include at least one number character

The password policy:
+ Include 16 characters (Counted Null Character)
+ Must include at least one low case character
+ Must include at least one up case character
+ Must include at least one number character
+ Must include at least one special character

Tested:
1. Normal testing:
```
$ipmitool raw 0x2c 0x2 0x52 0xa5
 52 5a 49 6c 43 4a 35 5a 4a 4a 30 72 36 32 36 61
 34 00 34 6d 68 4b 66 6f 37 24 73 34 68 38 23 52
 65 35 00
$ipmitool raw 0x2c 0x2 0x52 0xa1
 52 6f 6d 38 64 35 41 72 6c 69 72 67 78 35 37 37
 69 00 32 38 23 31 63 50 78 24 68 41 6c 30 57 70
 23 32 00
```
2. Create the 16th bootStrap account should be failed.
```
$ipmitool raw 0x2c 0x2 0x52 0xa1
Unable to send RAW command (channel=0x0 netfn=0x2c lun=0x0 cmd=0x2
rsp=0x80): Unknown (0x80)
$ipmitool raw 0x2c 0x1 0x52 0x1
Unable to send RAW command (channel=0x0 netfn=0x2c lun=0x0 cmd=0x1
rsp=0x80): Unknown (0x80)
```
3. Remove one bootStrap account use Web/Redfish.
4. Create new bootStrap account should be passed.

Upstream-Status: Pending

Change-Id: I02865af73f8108c07799b1539fe80294bf7fe291
Signed-off-by: Thu Nguyen <thu@os.amperecomputing.com>
---
 redfishhostinterface.cpp   | 496 +++++++++++++++++++++++++++++++++++++
 redfishhostinterface.hpp   |  14 ++
 user_channel/user_mgmt.hpp |   1 +
 3 files changed, 511 insertions(+)

diff --git a/redfishhostinterface.cpp b/redfishhostinterface.cpp
index 00d19d0..105d226 100644
--- a/redfishhostinterface.cpp
+++ b/redfishhostinterface.cpp
@@ -32,6 +32,8 @@
 #include <user_channel/user_layer.hpp>
 #include <user_channel/user_mgmt.hpp>
 
+#include <ranges>
+
 using namespace phosphor::logging;
 
 static constexpr const char* credBootstrappingInf =
@@ -201,12 +203,506 @@ ipmi::Cc filterRHICommands(ipmi::message::Request::ptr request)
     return ipmi::ccSuccess;
 }
 
+/** @brief Check whether the user name is valid
+ *
+ *  @param[in] userName - user name which needs validate.
+ *
+ *  @returns true when userName is valid, otherwise return false
+ */
+bool validUserName(const std::string& useName)
+{
+    bool haveUpChar = false;
+    bool haveLowChar = false;
+    std::string lowCharacters = static_cast<std::string>(ipmi::lowCharacters);
+    std::string upCharacters = static_cast<std::string>(ipmi::upCharacters);
+    std::string numberCharacters =
+        static_cast<std::string>(ipmi::numberCharacters);
+
+    for (auto& useNameChar : useName)
+    {
+        if (lowCharacters.find(useNameChar) != std::string::npos)
+        {
+            haveLowChar = true;
+            continue;
+        }
+        if (upCharacters.find(useNameChar) != std::string::npos)
+        {
+            haveUpChar = true;
+            continue;
+        }
+    }
+
+    if (haveLowChar && haveUpChar)
+    {
+        return true;
+    }
+
+    return false;
+}
+
+/** @brief Create random BootStrap Account Name
+ *
+ *  @param[in] userNameLength - user name length
+ *
+ *  @returns one random BootStrap Account Name
+ */
+std::string createRandomUserName(const uint8_t& userNamelength)
+{
+    /*
+     * Create random username has 16 characters and follow this policy
+     * + Include 16 characters
+     * + Must include at least one Low case character
+     * + Must include at least one Up case character
+     * + Must include at least one number character
+     */
+    std::string userName = "";
+    std::string lowCharacters = static_cast<std::string>(ipmi::lowCharacters);
+    std::string upCharacters = static_cast<std::string>(ipmi::upCharacters);
+    std::string numberCharacters =
+        static_cast<std::string>(ipmi::numberCharacters);
+    std::vector<unsigned char> randomCharType(userNamelength);
+    std::vector<unsigned char> randomChar(userNamelength);
+
+    if (RAND_bytes(randomCharType.data(), userNamelength) != 1)
+    {
+        lg2::error("Error generating random bytes with OpenSSL.");
+        return userName;
+    }
+
+    if (RAND_bytes(randomChar.data(), userNamelength) != 1)
+    {
+        // Handle error if RAND_bytes fails
+        lg2::error("Error generating random bytes with OpenSSL.");
+        return userName;
+    }
+
+    for (auto i = 0; i < userNamelength; i++)
+    {
+        /*
+         * User name format [a-zA-Z_][a-zA-Z_0-9]*
+         * First charracter should be low or up case
+         */
+        auto type = randomCharType[i] % ipmi::userNameCharacterTypes;
+        if (i == 0)
+        {
+            type = randomCharType[i] % 2;
+        }
+
+        switch (type)
+        {
+            case 0:
+                userName +=
+                    lowCharacters[randomChar[i] % lowCharacters.length()];
+                break;
+            case 1:
+                userName += upCharacters[randomChar[i] % upCharacters.length()];
+                break;
+            case 2:
+                userName +=
+                    numberCharacters[randomChar[i] % numberCharacters.length()];
+                break;
+            default:
+                break;
+        }
+    }
+    return userName;
+}
+
+/** @brief Create BootStrap Account Name
+ *
+ *  @param[in] userNameLength - user name length
+ *
+ *  @returns one valid BootStrap Account Name
+ */
+std::string createUserName(const uint8_t& userNameLength)
+{
+    std::string userName;
+    for ([[maybe_unused]] auto i : std::views::iota(
+             0, static_cast<int>(ipmi::maxCreateRandomUserNameRetry)))
+    {
+        userName = createRandomUserName(userNameLength);
+        if (userName.empty())
+        {
+            lg2::error("Failed to create BootStrap userName.");
+            continue;
+        }
+        if (validUserName(userName))
+        {
+            return userName;
+        }
+    }
+
+    return "";
+}
+
+/** @brief Check whether the password follow the common policy
+ *
+ *  @param[in] password - password.
+ *
+ *  @returns true when password is valid, otherwise return false
+ */
+bool validPassword(const std::string& password)
+{
+    bool haveUpChar = false;
+    bool haveLowChar = false;
+    bool haveNumberChar = false;
+    bool haveSpecialChar = false;
+    std::string lowCharacters = static_cast<std::string>(ipmi::lowCharacters);
+    std::string upCharacters = static_cast<std::string>(ipmi::upCharacters);
+    std::string numberCharacters =
+        static_cast<std::string>(ipmi::numberCharacters);
+    std::string specialCharacters =
+        static_cast<std::string>(ipmi::specialCharacters);
+
+    for (auto& passwordChar : password)
+    {
+        if (lowCharacters.find(passwordChar) != std::string::npos)
+        {
+            haveLowChar = true;
+            continue;
+        }
+        if (upCharacters.find(passwordChar) != std::string::npos)
+        {
+            haveUpChar = true;
+            continue;
+        }
+        if (numberCharacters.find(passwordChar) != std::string::npos)
+        {
+            haveNumberChar = true;
+            continue;
+        }
+        if (specialCharacters.find(passwordChar) != std::string::npos)
+        {
+            haveSpecialChar = true;
+            continue;
+        }
+    }
+
+    if (haveLowChar && haveUpChar && haveNumberChar && haveSpecialChar)
+    {
+        return true;
+    }
+
+    return false;
+}
+
+/** @brief Create one random user password
+ *
+ *  @param[in] passwordLength - length of the random password.
+ *
+ *  @returns one random password
+ */
+std::string createRandomPassword(const uint8_t& passwordLength)
+{
+    /*
+     * Create random password has 16 characters and follow common policy
+     * + Include 16 characters
+     * + Must include at least one Low case character
+     * + Must include at least one Up case character
+     * + Must include at least one number character
+     * + Must include at least one special character
+     */
+    std::string password = "";
+    std::string lowCharacters = static_cast<std::string>(ipmi::lowCharacters);
+    std::string upCharacters = static_cast<std::string>(ipmi::upCharacters);
+    std::string numberCharacters =
+        static_cast<std::string>(ipmi::numberCharacters);
+    std::string specialCharacters =
+        static_cast<std::string>(ipmi::specialCharacters);
+    std::vector<unsigned char> randomCharType(passwordLength);
+    std::vector<unsigned char> randomChar(passwordLength);
+
+    if (RAND_bytes(randomCharType.data(), passwordLength) != 1)
+    {
+        lg2::error("Error generating random bytes with OpenSSL.");
+        return password;
+    }
+
+    if (RAND_bytes(randomChar.data(), passwordLength) != 1)
+    {
+        // Handle error if RAND_bytes fails
+        lg2::error("Error generating random bytes with OpenSSL.");
+        return password;
+    }
+
+    for (auto i = 0; i < passwordLength; i++)
+    {
+        auto type = randomCharType[i] % ipmi::passwordCharacterTypes;
+        switch (type)
+        {
+            case 0:
+                password +=
+                    lowCharacters[randomChar[i] % lowCharacters.length()];
+                break;
+            case 1:
+                password += upCharacters[randomChar[i] % upCharacters.length()];
+                break;
+            case 2:
+                password +=
+                    numberCharacters[randomChar[i] % numberCharacters.length()];
+                break;
+            case 3:
+                password += specialCharacters[randomChar[i] %
+                                              specialCharacters.length()];
+                break;
+            default:
+                break;
+        }
+    }
+    return password;
+}
+
+/** @brief Create one valid password
+ *
+ *  @param[in] passwordLength - length of the valid password.
+ *
+ *  @returns password
+ */
+std::string createPassword(const uint8_t& passwordLength)
+{
+    std::string password;
+    for ([[maybe_unused]] auto i : std::views::iota(
+             0, static_cast<int>(ipmi::maxCreateRandomPassworkRetry)))
+    {
+        password = createRandomPassword(passwordLength);
+        if (password.empty())
+        {
+            lg2::error("Failed to create password for BootStrap account.");
+            continue;
+        }
+        if (validPassword(password))
+        {
+            return password;
+        }
+    }
+
+    return "";
+}
+
+/** @brief implements the set CredentialBootstrapping's Enabled property
+ *
+ *  @param[in] pValue - Context pointer
+ *  @param[in] pValue - enable status
+ *
+ *  @returns none
+ */
+void setCredentialBootstrapEnabledProperty(ipmi::Context::ptr ctx,
+                                           const bool& pValue)
+{
+    std::string service;
+
+    try
+    {
+        ipmi::setDbusProperty(*ctx->bus, ipmi::userMgrService,
+                              ipmi::userObjBasePath, credBootstrappingInf,
+                              credBootstrapEnabledProp, pValue);
+    }
+    catch (const std::exception& e)
+    {
+        lg2::error(
+            "Can't set value of {PROP} property in the {INF} interface at path {PATH} error {ERROR}",
+            "PROP", credBootstrapEnabledProp, "INF", credBootstrappingInf,
+            "PATH", ipmi::userObjBasePath, "ERROR", e);
+    }
+}
+
+/** @brief Implements the Get bootstrap account credentials command
+ *
+ *  @param[in] ctx - shared_ptr to an IPMI context struct
+ *  @param[in] bootstrapControl - Disable credential bootstrapping control
+ *
+ *  @returns IPMI completion code plus response data for
+ *   - the Get bootstrap account credentials command
+ */
+ipmi::RspType<std::vector<uint8_t>, /* 16 bytes user name */
+              std::vector<uint8_t>> /* 16 bytes user password */
+    getBootstrapAccountCredentials(ipmi::Context::ptr ctx,
+                                   uint8_t bootstrapControl)
+{
+    ipmi::ChannelInfo chInfo;
+    std::string userName = "";
+    std::string password;
+    uint8_t userId = 1;
+    ipmi::PrivAccess privAccess = {};
+    ipmi::UsersTbl* userData;
+    uint8_t userCnt = 1; /* One Reserve user account */
+
+    if (ipmi::getChannelInfo(ctx->channel, chInfo) != ipmi::ccSuccess)
+    {
+        lg2::error("Failed to get Channel Info, channel={CHANNEL}", "CHANNEL",
+                   ctx->channel);
+        return ipmi::responseUnspecifiedError();
+    }
+
+    if (!checkRHIAllowedMediumType(chInfo.mediumType))
+    {
+        lg2::error("Error - Medium interface not supported, medium={TYPE}",
+                   "TYPE", chInfo.mediumType);
+        return ipmi::responseCommandNotAvailable();
+    }
+
+    /* Get Enabled property within the CredentialBootstrapping property of the
+     * host interface. */
+    auto enabledState = getCredentialBootstrapEnabledProperty(ctx);
+    if (!enabledState || (*enabledState == false))
+    {
+        lg2::error("CredentialBootstrapping is disabled.");
+        return ipmi::responseCmdDisabled();
+    }
+
+    /* Find the user ID empty to set the user name */
+    userData = ipmi::getUserAccessObject().getUsersTblPtr();
+    for (uint8_t usrIndex = 1; usrIndex <= ipmi::ipmiMaxUsers; ++usrIndex)
+    {
+        if (userData->user[usrIndex].userInSystem)
+        {
+            userCnt++;
+            continue;
+        }
+
+        userId = usrIndex;
+        break;
+    }
+    if (userCnt >= ipmi::ipmiMaxUsers)
+    {
+        lg2::error("Invalid User ID - Out of range");
+        return ipmi::responseParmOutOfRange();
+    }
+
+    /*
+     * As DSP0270:
+     * 8.1.2 Get bootstrap account credentials (NetFn 2Ch, Command 02h)
+     * The response data bytes[3:18]:
+     * "The user name as a UTF-8 string. Strings with fewer than 16 characters
+     * are terminated with a null (00h) character and 00h padded to 16 bytes.
+     * => User name length without null (00h) is 15 bytes.
+     */
+    userName = createUserName(ipmi::ipmiMaxUserName - 1);
+    if (userName.empty())
+    {
+        lg2::error("Failed to create BootStrap userName.");
+        return ipmi::responseUnspecifiedError();
+    }
+
+    if (ipmi::ipmiUserAddUserToNonIpmiGroupUsers(userName) != ipmi::ccSuccess)
+    {
+        lg2::error("Non-ipmi User limit reached");
+        return ipmi::responseOutOfSpace();
+    }
+
+    auto rc = ipmi::ipmiUserSetUserName(userId, userName);
+    if (rc != ipmi::ccSuccess)
+    {
+        lg2::error(
+            "Failed to set User name userName {NAME} error code {ERROR}.",
+            "NAME", userName, "ERROR", rc);
+        return ipmi::responseUnspecifiedError();
+    }
+
+    for ([[maybe_unused]] auto i :
+         std::views::iota(0, static_cast<int>(ipmi::maxIpmiUserActionRetry)))
+    {
+        /*
+         * As DSP0270:
+         * 8.1.2 Get bootstrap account credentials (NetFn 2Ch, Command 02h)
+         * The response data bytes[19:34]:
+         * "The password as a UTF-8 string. Strings with fewer than 16
+         * characters are terminated with a null (00h) character and 00h padded
+         * to 16 bytes.
+         * => Password length without null (00h) is 15 bytes.
+         */
+        password = createPassword(ipmi::maxIpmi15PasswordSize - 1);
+        if (password.empty())
+        {
+            lg2::error("Failed to create password for BootStrap account.");
+            continue;
+        }
+        rc = ipmi::ipmiUserSetUserPassword(userId, password.c_str());
+        if (rc == ipmi::ccSuccess)
+        {
+            break;
+        }
+    }
+
+    if (rc != ipmi::ccSuccess)
+    {
+        lg2::error(
+            "Failed to set UserPassword userName {NAME} error code {ERROR}.",
+            "NAME", userName, "ERROR", rc);
+        return ipmi::responseUnspecifiedError();
+    }
+
+    /* Set user privilege for bootStrap account is User */
+    privAccess.privilege = PRIVILEGE_USER;
+    rc = ipmi::ipmiUserSetPrivilegeAccess(static_cast<uint8_t>(userId),
+                                          ctx->channel, privAccess, 0);
+    if (rc != ipmi::ccSuccess)
+    {
+        lg2::error(
+            "Failed to set User Privilege Access userName {NAME} error code {ERROR}.",
+            "NAME", userName, "ERROR", rc);
+        return ipmi::responseUnspecifiedError();
+    }
+
+    auto userGroup = {static_cast<std::string>(ipmi::redfishGrpName)};
+    /* Set UserGroup of bootStrap account to {"redfish"} */
+    rc = ipmi::ipmiUserSetUserGroups(userId, ctx->channel, userGroup);
+    if (rc != ipmi::ccSuccess)
+    {
+        lg2::error("Failed to create BootStrap userName.");
+        return ipmi::responseUnspecifiedError();
+    }
+
+    /* Set IsBootStrap of bootStrap account to true */
+    rc = ipmi::ipmiUserSetUserBootStrapAccountState(userId, true);
+    if (rc != ipmi::ccSuccess)
+    {
+        lg2::error(
+            "Failed to set User IsBootStrap property userName {NAME} error code {ERROR}.",
+            "NAME", userName, "ERROR", rc);
+        return ipmi::responseUnspecifiedError();
+    }
+
+    /* Enable user */
+    rc = ipmi::ipmiUserUpdateEnabledState(userId, true);
+    if (rc != ipmi::ccSuccess)
+    {
+        lg2::error("Failed to enable userName {NAME} error code {ERROR}.",
+                   "NAME", userName, "ERROR", rc);
+        return ipmi::responseUnspecifiedError();
+    }
+
+    if (bootstrapControl != ipmi::keepCredBootstrapEnabled)
+    {
+        lg2::info("Set CredentialBootstrapping to disabled");
+        setCredentialBootstrapEnabledProperty(ctx, false);
+    }
+    else
+    {
+        lg2::info("Keep credential bootstrapping enabled");
+    }
+
+    /* Respond data */
+    std::vector<uint8_t> vUserName;
+    vUserName.insert(vUserName.end(), userName.begin(), userName.end());
+    vUserName.push_back('\0');
+    std::vector<uint8_t> vPassword;
+    vPassword.insert(vPassword.end(), password.begin(), password.end());
+    vPassword.push_back('\0');
+
+    return ipmi::responseSuccess(vUserName, vPassword);
+}
+
 void registerNetfnRedfishHostInterfaceFunctions() __attribute__((constructor));
 void registerNetfnRedfishHostInterfaceFunctions()
 {
     ipmi::registerGroupHandler(ipmi::prioOpenBmcBase, ipmi::groupRedfish,
                                ipmi::bootStrap::cmdGetMngCertFingerprint,
                                ipmi::Privilege::User, getFingerprint);
+    ipmi::registerGroupHandler(
+        ipmi::prioOpenBmcBase, ipmi::groupRedfish,
+        ipmi::bootStrap::cmdGetBootstrapAccoutCre, ipmi::Privilege::User,
+        getBootstrapAccountCredentials);
     ipmi::registerFilter(ipmi::prioOpenBmcBase,
                          [](ipmi::message::Request::ptr request) {
                              return filterRHICommands(request);
diff --git a/redfishhostinterface.hpp b/redfishhostinterface.hpp
index 6e5ade9..b9d3eef 100644
--- a/redfishhostinterface.hpp
+++ b/redfishhostinterface.hpp
@@ -52,4 +52,18 @@ static inline auto responseCertsNumberInvalid()
     return response(ccCertificateNumberInvalid);
 }
 
+/* Retry in ipmi user actions */
+static constexpr uint8_t maxIpmiUserActionRetry = 5;
+
+/* Retry in creating a valid user and password */
+static constexpr uint8_t maxCreateRandomPassworkRetry = 10;
+static constexpr uint8_t maxCreateRandomUserNameRetry = 10;
+
+static constexpr uint8_t passwordCharacterTypes = 4;
+static constexpr uint8_t userNameCharacterTypes = 3;
+static constexpr const char* lowCharacters = "abcdefghijklmnopqrstuvwxyz";
+static constexpr const char* upCharacters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+static constexpr const char* numberCharacters = "0123456789";
+static constexpr const char* specialCharacters = "!@#$%^&*";
+
 } // namespace ipmi
diff --git a/user_channel/user_mgmt.hpp b/user_channel/user_mgmt.hpp
index fc62960..daa18f0 100644
--- a/user_channel/user_mgmt.hpp
+++ b/user_channel/user_mgmt.hpp
@@ -49,6 +49,7 @@ static constexpr const char* ipmiUserSignalLockFile =
     "/run/ipmi/ipmi_usr_signal_mutex";
 static constexpr const char* ipmiUserDataFile = "/var/lib/ipmi/ipmi_user.json";
 static constexpr const char* ipmiGrpName = "ipmi";
+static constexpr const char* redfishGrpName = "redfish";
 static constexpr size_t privNoAccess = 0xF;
 static constexpr size_t privMask = 0xF;
 
-- 
2.34.1

